base/fastmath.jl

```diff
@@ -0,0 +1,326 @@
+            r = rem(x,y)
+            ifelse((r > 0) $ (y > 0), r+y, r)
+        end
+    end
+end
+
+isinf_fast(x) = false
+isfinite_fast(x) = true
+isnan_fast(x) = false
+issubnormal_fast(x) = false
+
+# complex numbers
+
+div_fast(x::Complex64, y::Float32) = @fastmath Complex64(real(x)/y, imag(x)/y)
+div_fast(x::Complex128, y::Float64) = @fastmath Complex128(real(x)/y, imag(x)/y)
+
+@fastmath for T in (Complex64, Complex128)
+    @eval begin
+        abs_fast(x::$T) = hypot(real(x), imag(x))
+        abs2_fast(x::$T) = real(x)*real(x) + imag(x)*imag(x)
+        add_fast(x::$T, y::$T) = $T(real(x)+real(y), imag(x)+imag(y))
+        conj_fast(x::$T) = $T(real(x), -imag(x))
+        function div_fast(x::$T, y::$T)
+            $T(real(x)*real(y) + imag(x)*imag(y),
+               imag(x)*real(y) - real(x)*imag(y)) / abs2(y)
+        end
+        inv_fast(x::$T) = conj(x) / abs2(x)
+        sign_fast(x::$T) = x / abs(x)
+        sub_fast(x::$T, y::$T) = $T(real(x)-real(y), imag(x)-imag(y))
+        function mul_fast(x::$T, y::$T)
+            $T(real(x)*real(y) - imag(x)*imag(y),
+               real(x)*imag(y) + imag(x)*real(y))
+        end
+
+        eq_fast(x::$T, y::$T) = (real(x)==real(y)) & (imag(x)==imag(y))
+        ne_fast(x::$T, y::$T) = !(x==y)
+    end
+end
+
+# fall-back implementations and type promotion
+
+for op in (:+, :-, :*, :/, :(==), :!=, :<, :<=,
+           :abs, :abs2, :cmp, :conj, :inv, :mod, :rem, :sign)
+    op_fast = fast_op[op]
+    @eval begin
+        # fall-back implementation for non-numeric types
+        ($op_fast)(xs...) = ($op)(xs...)
+        # type promotion
+        ($op_fast)(x::Number, y::Number, zs::Number...) =
+            ($op_fast)(promote(x,y,zs...)...)
+        # fall-back implementation that applies after promotion
+        ($op_fast){T<:Number}(x::T,ys::T...) = ($op)(x,ys...)
+    end
+end
+
+
+
+# Math functions
+
+# builtins
+
+pow_fast(x::Float32, y::Integer) =
+    box(Float32, Base.powi_llvm(unbox(Float32,x), unbox(Int32,int32(y))))
+pow_fast(x::Float64, y::Integer) =
+    box(Float64, Base.powi_llvm(unbox(Float64,x), unbox(Int32,int32(y))))
+
+# TODO: Change sqrt_llvm intrinsic to avoid nan checking; add nan
+# checking to sqrt in math.jl; remove sqrt_llvm_fast intrinsic
+sqrt_fast(x::Float32) = box(Float32, Base.sqrt_llvm_fast(unbox(Float32,x)))
+sqrt_fast(x::Float64) = box(Float64, Base.sqrt_llvm_fast(unbox(Float64,x)))
+
+# libm
+
+const libm = Base.libm_name
+
+for f in (:acos, :acosh, :asin, :asinh, :atan, :atanh, :cbrt, :cos,
+          :cosh, :exp2, :exp, :expm1, :lgamma, :log10, :log1p, :log2,
+          :log, :sin, :sinh, :tan, :tanh)
+    f_fast = fast_op[f]
+    @eval begin
+        ($f_fast)(x::Float32) =
+            ccall(($(string(f,"f")),libm), Float32, (Float32,), x)
+        ($f_fast)(x::Float64) =
+            ccall(($(string(f)),libm), Float64, (Float64,), x)
+    end
+end
+
+for f in (:^, :atan2)
+    f_fast = fast_op[f]
+    if f==:^; f=:pow; end       # special case
```

Comment:
With just two methods it would probably be better to just write them out explicitly.